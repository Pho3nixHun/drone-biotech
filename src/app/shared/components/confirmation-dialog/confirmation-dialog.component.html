@if (vm(); as vm) {
    <app-dialog-layout
        *transloco="let t"
        class="w-80 md:w-96 block"
        (close)="onCancelClick()"
        [vm]="vm"
    >
        <p>{{ t(vm.confirmTextKey) }}</p>
        <button (click)="onConfirmClick()" class="btn btn-primary">
            {{ t(vm.confirmButtonTextKey) }}
        </button>
        <button (click)="onCancelClick()" class="btn btn-secondary">
            {{ t(vm.cancelButtonTextKey) }}
        </button>
    </app-dialog-layout>
}
