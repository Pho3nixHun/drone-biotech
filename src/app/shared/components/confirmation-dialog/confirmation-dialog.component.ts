import { Component, ElementRef, input, output, viewChild } from '@angular/core';
import { DialogLayoutComponent } from '@components/dialog-layout/dialog-layout.component';
import { TranslocoModule } from '@jsverse/transloco';
import { ButtonComponent } from '@components/button/button.component';
import {
    ConfirmationDialogResponse,
    ConfirmationDialogVM,
} from './confirmation-dialog.model';
import { MatIconModule } from '@angular/material/icon';

@Component({
    selector: 'app-confirmation-dialog',
    imports: [
        DialogLayoutComponent,
        TranslocoModule,
        ButtonComponent,
        MatIconModule,
    ],
    templateUrl: './confirmation-dialog.component.html',
})
export class ConfirmationDialogComponent {
    public readonly vm = input.required<ConfirmationDialogVM>();
    public readonly response = output<ConfirmationDialogResponse>();
    public readonly dialog =
        viewChild.required<ElementRef<HTMLDialogElement>>('dialog');

    protected cancelClick() {
        this.response.emit({ type: 'cancel' });
        this.dialog().nativeElement.close();
    }
    protected submitClick() {
        this.response.emit({ type: 'submit' });
        this.dialog().nativeElement.close();
    }

    public openDialog() {
        this.dialog().nativeElement.showModal();
    }
}
