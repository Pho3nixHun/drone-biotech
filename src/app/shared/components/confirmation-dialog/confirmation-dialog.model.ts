import { ButtonVM } from '@components/button/button.model';
import { DialogLayoutVM } from '@components/dialog-layout/dialog-layout.model';
import { DialogReason } from '@services/dialog/dialog.service';
import { isObject } from '@utils/is-object.typeguard';

export interface ConfirmationDialogVM extends DialogLayoutVM {
    type: 'confirmationDialogVM';
    confirmTextKey: string;
    confirmButtonVM: ButtonVM;
    cancelButtonVM: ButtonVM;
}

export interface SubmitReason extends DialogReason {
    reasonType: 'submit';
}
export interface CancelReason extends DialogReason {
    reasonType: 'cancel';
}

export type ConfirmationDialogReason = SubmitReason | CancelReason;

export const isConfirmationDialogVM = (
    vm: unknown
): vm is ConfirmationDialogVM =>
    isObject(vm) && 'type' in vm && vm.type === 'confirmationDialogVM';
