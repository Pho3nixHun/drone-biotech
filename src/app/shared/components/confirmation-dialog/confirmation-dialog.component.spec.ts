import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ConfirmationDialogComponent } from './confirmation-dialog.component';
import { ConfirmationDialogVM } from './confirmation-dialog.model';
import { getTranslocoModule } from 'transloco-testing.module';
import { MatIcon } from '@interfaces/mat-icon.enum';

describe('ConfirmationDialogComponent', () => {
    let fixture: ComponentFixture<ConfirmationDialogComponent>;
    let compiled: HTMLElement;
    let component: ConfirmationDialogComponent;

    beforeEach(async () => {
        HTMLDialogElement.prototype.showModal = jest.fn(function () {
            this.setAttribute('open', '');
        });
        HTMLDialogElement.prototype.close = jest.fn(function () {
            this.removeAttribute('open');
        });
        await TestBed.configureTestingModule({
            imports: [
                ConfirmationDialogComponent,
                getTranslocoModule({
                    langs: { en: enMock },
                    translocoConfig: {
                        availableLangs: ['en'],
                        defaultLang: 'en',
                    },
                }),
            ],
        }).compileComponents();
        fixture = TestBed.createComponent(ConfirmationDialogComponent);
        component = fixture.componentInstance;
        compiled = fixture.debugElement.nativeElement;
    });

    // Snapshot testing
    it('should render the template correctly ', () => {
        // Arrange
        fixture.componentRef.setInput('vm', mockVM);

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });

    // Snapshot testing
    it('should render the template correctly if the dialog is opened', () => {
        // Arrange
        fixture.componentRef.setInput('vm', mockVM);

        // Act
        fixture.detectChanges();
        component['openDialog']();

        // Assert
        expect(compiled).toMatchSnapshot();
    });

    // Snapshot testing
    it('should close the dialog and send cancel response', () => {
        // Arrange
        const responseSpy = jest.spyOn(component['response'], 'emit');
        fixture.componentRef.setInput('vm', mockVM);

        // Act
        fixture.detectChanges();
        component['cancelClick']();

        // Assert
        expect(responseSpy).toHaveBeenCalledWith({ type: 'cancel' });
        expect(compiled).toMatchSnapshot();
    });

    // Snapshot testing
    it('should close the dialog and send submit response', () => {
        // Arrange
        const responseSpy = jest.spyOn(component['response'], 'emit');
        fixture.componentRef.setInput('vm', mockVM);

        // Act
        fixture.detectChanges();
        component['submitClick']();

        // Assert
        expect(responseSpy).toHaveBeenCalledWith({ type: 'submit' });
        expect(compiled).toMatchSnapshot();
    });
});

const enMock = {
    title: 'tit',
    confirmText: 'confirm',
};

const mockVM: ConfirmationDialogVM = {
    titleKey: enMock.title,
    confirmTextKey: enMock.confirmText,
    closeButtonXVM: { icon: MatIcon.ADD, variant: 'fill' },
    confirmButtonXVM: { icon: MatIcon.ADD, variant: 'fill' },
    cancelButtonXVM: {
        icon: MatIcon.ADD,
        variant: 'fill',
    },
};
