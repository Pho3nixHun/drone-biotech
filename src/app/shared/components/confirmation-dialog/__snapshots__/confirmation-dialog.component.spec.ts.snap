// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ConfirmationDialogComponent should not render the template if not the correct vm is provided 1`] = `
<div
  id="root0"
  ng-version="20.1.7"
>
  <!--container-->
</div>
`;

exports[`ConfirmationDialogComponent should render the template if the correct vm is provided 1`] = `
<div
  id="root1"
  ng-version="20.1.7"
>
  <app-dialog-layout
    class="w-80 md:w-96 block"
  >
    <section
      class="p-4 md:p-10 rounded-lg shadow-md transition-all duration-100 bg-white space-y-5 @container/dialog"
    >
      <div
        class="flex justify-between"
      >
        <h3
          class="text-3xl"
        >
           tit 
        </h3>
        <button>
          <mat-icon
            aria-hidden="true"
            class="mat-icon notranslate text-3xl material-icons mat-ligature-font mat-icon-no-color"
            data-mat-icon-type="font"
            role="img"
            style="width: 40px; height: 40px; display: flex; justify-content: center; align-items: center;"
          >
            close
          </mat-icon>
        </button>
      </div>
      <div
        class="space-y-4"
      >
        <p>
          confirm
        </p>
      </div>
      <div
        class="flex justify-between @lg/dialog:justify-end @lg/dialog:space-x-3"
      >
        <button
          class="btn btn-primary"
        >
           confirmButton 
        </button>
        <button
          class="btn btn-secondary"
        >
           cancelButton 
        </button>
      </div>
    </section>
    <!--container-->
  </app-dialog-layout>
  <!--container-->
  <!--container-->
</div>
`;
